### PR TITLE
Update unit test to handle arbitrary token decimals

### LINKED ISSUE #735: Update unit test to handle arbitrary token decimals.
Update unit test to handle arbitrary token decimals.
Detail of the ticket is here https://github.com/OpenST/JLP/issues/98#issuecomment-492484237

### PR DESCRIPTION
**Pull request by schemar**
* Added config in tests to get configured decimals (default or `env` override).
* Updated `MockToken` to take decimals as constructor argument.
* Updated all instantiations of various tokens in unit tests to use decimals from config.
* Updated proof generator to use 18 decimals (it uses mock token).
* Added new script `test:range`. Example: `npm run test:range -- 16 18`

Fixes #735

### CODE REVIEW

**Comment by schemar** [comment]
Ready for reviow 🐉

File: tools/test_range.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+# Example run with decimals 12 - 18 (inclusive):
+#
+# ```bash
+# npm run test:range -- 12 18
+# ```
+###
+
+echo ""
+
+if [ $# -ne 2 ];
```
Comment (jasonklein) [inline tools/test_range.sh]: The acceptability criteria for this ticket indicate that entering 1 value should be acceptable.

Although, given your `seq` approach, I think including a statement in the echo documentation that providing the same value twice will achieve that result would be a simple way to satisfy the motivation behind this particular criterion.

File: tools/test_range.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+then
+    echo "Wrong number of parameters given."
+    echo "Run with two arguments: start and end of range inclusive."
+    echo "Example:"
+    echo "npm run test:range -- 16 20"
```
[the remark this answers is not in the document]
Comment (schemar) [reply tools/test_range.sh]: I don't know if it is required, but it is the [official way](https://docs.npmjs.com/cli/run-script):
`npm run-script <command> [--silent] [-- <args>...]`

Should I remove them if it works without? I only tested with.

File: tools/test_range.sh (hunk 1 above)
Comment (schemar) [reply tools/test_range.sh]: I will add that statement 👍

**Comment by schemar** [comment]
> * Should `MosaicCoreConfig` be updated?

I don't understand. Can you please elaborate?

> * Should the range script end with the first value to have failing tests (current behavior) or handle failing tests with each value in the range?

I will change the script to run all the tests and change the exit code based on whether a test fails.

File: tools/test_range.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasonklein) [reply tools/test_range.sh]: Ah! Thanks for the docs reference! I think it's better to leave the `--`. Please disregard my suggestion 🙏

**Comment by jasonklein** [comment]
> > * Should `MosaicCoreConfig` be updated?
> 
> I don't understand. Can you please elaborate?

Apologies for not being specific. From `MosaicCoreConfig`:
```
/** The number of decimals of the base token. */
uint256 public constant DECIMALSFACTOR = 10 ** uint256(18);
```
I asked because of the changes to base token references in this PR. On the other hand, perhaps those references shouldn't be dynamic and should always have 18 decimals?

**Comment by schemar** [comment]
> > > * Should `MosaicCoreConfig` be updated?
> > 
> > 
> > I don't understand. Can you please elaborate?
> 
> @schemar: apologies for not being specific. From `MosaicCoreConfig`:
> 
> ```
> /** The number of decimals of the base token. */
> uint256 public constant DECIMALSFACTOR = 10 ** uint256(18);
> ```
> 
> I asked because of the changes to base token references in this PR. On the other hand, perhaps those references shouldn't be dynamic and should always have 18 decimals?

Hm 🤔 Good point. Mosaic core is way outdated by now. It's probably not worth it to mingle around in the core code in the context of this PR.

File: tools/test_range.sh (hunk 1 above)
Comment (schemar) [reply tools/test_range.sh]: The script will accept one and two arguments :+1

**Comment by schemar** [comment]
Ready for review again. Now accepts only one argument and continues after failed decimal (prints all failed decimals at the end).

**Pull request merged**